### DISCUSSION TITLE
One. Crunching knowledge

### DISCUSSION THREAD

**Discussion by Mehdi-H** [Mehdi-H/DDD-blue-book-collaborative-reading#1] [Chapter]
# One. Crunching knowledge

You can discuss here about this chapter ! 👇

- Ask questions,
- or share anything you find insightful !

**Comment by sofcalca** [comment]
What I learned from this chapter :

## Characteristics of a good model:
- The model of the business process and its software implementation **are linked and evolve together**.
- It provides a common language between developers and business: the model and its language is understood by both parties and can be used to communicate without ambiguity.
- It is a way to **concentrate business knowledge**: it encapsulates the complexity of the business.
- It **evolves over time**: new concepts are added to it during its lifecycle, but obsolete or unnecessary concepts are also removed and refactored.
    - it starts out rather superficial, and becomes increasingly complex as business knowledge is enriched.
- **It emerges from discussions and brainstorming**. Discussions help to spot hot spots or odd things in the model

## Requirements
Domain-driven design (DDD) requires **short feedback loops** to enrich the team's knowledge as they learn.

- **Incompatible with waterfall**, where modeling is done at the beginning, often not by the people who will develop the solution.
- **Developers cannot be simple executors** who do not ask questions about the context of their software.
- **Developers cannot be isolated from business** : the model would then be too superficial to be truly impactful.

The model allows centralization of business knowledge from disparate sources: business, documentation, users, developers who have worked on existing systems...

## Effects of DDD

- Forces developers to understand the business.
- Code could be directly understood by the business

**Reply by Mehdi-H** [reply]
> Developers cannot be simple executors who do not ask questions about the context of their software.

This statement makes me think about [this episode of Café Craft podcast](https://open.spotify.com/episode/4IYFq6QAoQ91hBrrEJaiyE?si=fcfbca550bed401f) with [Christophe Thibault](https://twitter.com/ToF_) about Quality in software engineering, 

Quality cannot be achieved by "just doing", as simple executors, but by creating the conditions for the developers to learn : learn software craftsmanship practices like [TDD](https://martinfowler.com/bliki/TestDrivenDevelopment.html), but also learn about the domain, as software _designers_

**Comment by Mehdi-H** [comment]
In this chapter, I really liked the example of a "realistic discussion" between a software engineer and a [SME](https://en.wikipedia.org/wiki/Subject-matter_expert). I'm often frustrated by the injunction to "just discuss with people to understand the business" that often comes without guidelines or examples.

There are similar examples of such discussions in Scott Wlaschin's book [Domain Modeling Made Functional: Tackle Software Complexity with Domain-Driven Design and F#](https://fsharpforfunandprofit.com/books/#domain-modeling-made-functional) with concrete examples of how to turn those disucssions into a (functional and expressive) domain model.

**Reply by sofcalca** [reply]
Having being part of discussions between SMEs and developers, current tools that come to my mind are :
- [Event storming](https://www.eventstorming.com/)
- [Example mapping](https://cucumber.io/blog/bdd/example-mapping-introduction/)
But I am not sure the level of details from the discussion you are referring to can be achieved with these tools

**Comment by Mehdi-H** [comment]
Maybe I am projecting a bit much on personal experiences ( 😄 ) but one message I understand [from this chapter](https://learning.oreilly.com/library/view/domain-driven-design-tackling/0321125215/ch01.html#:-:text=These%20self-educated%20team%20,%20the%20most%20critical%20areas.) also is that building a rich domain model asks a lot of time and energy, and for the DDD approach to be effective that **one organization should not consider assigning its developers on multiple developement projects at a time**

**Reply by sofcalca** [reply]
I agree, and I think it is related to these points I quoted above:
> Developers cannot be simple executors who do not ask questions about the context of their software.
> Developers cannot be isolated from business : the model would then be too superficial to be truly impactful.

If developer have to code **and** have a good understanding of the domain, making them navigate between very different domains and context switch is costly.
